### PR TITLE
Add iputils dynamic application build from source

### PR DESCRIPTION
**Pull request by radupascale**

### CODE REVIEW

File: iputils_srcs/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+## Dependencies
```
Comment (StefanJum) [inline iputils_srcs/README.md:1]: Please add more details on how to run the application. You can use [this README file](https://github.com/unikraft/dynamic-apps/blob/bf3a5a2febccb0a8f507fe0b70e7a1db48cbbe3b/lang/lua/README.md?plain=1) for reference.

File: iputils_srcs/build.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+rm -rf iputils
+
+# echo "Downloading the source code"
+git clone https://github.com/iputils/iputils iputils
+
+# echo "Configuring..."
+cd iputils
```
Comment (StefanJum) [inline iputils_srcs/build.sh:9]:
```suggestion
pushd iputils &> /dev/null || exit 1
```

File: iputils_srcs/build.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+./configure && meson build
+
+# echo "Installing the build..."
+cd builddir && meson install
```
Comment (StefanJum) [inline iputils_srcs/build.sh:13]:
```suggestion
pushd builddir &> /dev/null || exit 1
meson install
```

File: iputils_srcs/build.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+# echo "Copying ping, tracepath, clockdiff, arping..."
+mkdir -p ../../bin
+cp ./arping ./clockdiff ./tracepath ./ping/ping ../../bin/
+
+cd ../..
```
Comment (StefanJum) [inline iputils_srcs/build.sh:19]:
```suggestion
popd &> /dev/null
```
You might need to do `popd` 2 times here.

**Review by StefanJum — CHANGES_REQUESTED** [review]
Looks good overall @radupascale. See the comments.
Also after you create the script, use [shellcheck](https://www.shellcheck.net/) to test it.

File: iputils_srcs/README.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+```
```
Comment (StefanJum) [inline iputils_srcs/README.md:2]:
```suggestion
```console
```

File: iputils_srcs/build.sh [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+# echo "Installing the build..."
+pushd builddir &> /dev/null || exit 1
+meson install
+
+# echo "Copying ping, tracepath, clockdiff, arping..."
+mkdir -p ../../bin
+cp ./arping ./clockdiff ./tracepath ./ping/ping ../../bin/
+
+popd &> /dev/null
```
Comment (StefanJum) [inline iputils_srcs/build.sh:20]:
```suggestion
popd &> /dev/null || exit 1
```

File: iputils_srcs/build.sh [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+popd &> /dev/null
```
Comment (StefanJum) [inline iputils_srcs/build.sh:21]:
```suggestion
popd &> /dev/null || exit 1
```

File: iputils_srcs/build.sh [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+../extract.sh ./bin/tracepath
```
Comment (StefanJum) [inline iputils_srcs/build.sh:22]:
```suggestion

../extract.sh ./bin/tracepath
```

**Review by StefanJum — CHANGES_REQUESTED** [review]
For some reason I get this when building:
```
doc/meson.build:23:0: ERROR: Program(s) ['xsltproc'] not found or not executable

A full log can be found at /media/stefan/projects/unikraft/usoc23/session-04/dynamic-apps/iputils_srcs/iputils/builddir/meson-logs/meson-log.txt
Install data not found. Run this command in build directory root.
cp: cannot stat './arping': No such file or directory
cp: cannot stat './clockdiff': No such file or directory
cp: cannot stat './tracepath': No such file or directory
cp: cannot stat './ping/ping': No such file or directory
./bin/tracepath is not a file
./bin/tracepath is not a file
./bin/clockdiff is not a file
./bin/ping is not a file
```
Is there some dependency I'm missing?

File: iputils_srcs/build.sh (hunk 8 above)
Comment (radupascale) [reply iputils_srcs/build.sh:23]: Yeah, my bad `xsltproc` should have been listed as a dependency. I modified the README to include it as well. It should work now.

**Comment by radupascale** [comment]
> 



> For some reason I get this when building:
> 
> ```
> doc/meson.build:23:0: ERROR: Program(s) ['xsltproc'] not found or not executable
> 
> A full log can be found at /media/stefan/projects/unikraft/usoc23/session-04/dynamic-apps/iputils_srcs/iputils/builddir/meson-logs/meson-log.txt
> Install data not found. Run this command in build directory root.
> cp: cannot stat './arping': No such file or directory
> cp: cannot stat './clockdiff': No such file or directory
> cp: cannot stat './tracepath': No such file or directory
> cp: cannot stat './ping/ping': No such file or directory
> ./bin/tracepath is not a file
> ./bin/tracepath is not a file
> ./bin/clockdiff is not a file
> ./bin/ping is not a file
> ```
> 
> Is there some dependency I'm missing?

Yeah, my bad `xsltproc` should have been listed as a dependency. I modified the README to include it as well. It should work now.

File: iputils_srcs/README.md (hunk 5 above)
Comment (StefanJum) [inline iputils_srcs/README.md:2]:
```suggestion
```console
```

Comment (StefanJum) [reply iputils_srcs/README.md:2]: Add an empty line before and after the code snippets

File: iputils_srcs/README.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+sudo apt install meson libcap-dev xsltproc
+```
+## Installation
+```
```
Comment (StefanJum) [inline iputils_srcs/README.md:6]:
```suggestion
```console
```

Comment (StefanJum) [reply iputils_srcs/README.md:6]: Add an empty line before and after the code snippets

**Review by StefanJum — CHANGES_REQUESTED** [review]
Thanks @radupascale, the script works fine now :tada: 

I still need to run it as `sudo`, otherwise I'll get `PermissionError: [Errno 13] Permission denied: '/usr/local/share/locale/cs/LC_MESSAGES/iputils.mo.tmp'`. If this is how it's supposed to be run please update the README file.

Also see the few comments on the README file.

File: iputils-src/README.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+The script will download the [`iputils` repository](https://github.com/iputils/iputils), install the required dependencies and build the sources.
+
+## Run with Linux
+
+From inside the `app-iputils` directory, run:
+
+```console
+./bin/ping 8.8.8.8
+```
+
```
Comment (RaduNichita) [inline iputils-src/README.md:22]:
```suggestion
You should see the following output:

```

File: iputils-src/README.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+Do the same for the other applications (`arping`, `clockdiff`, `tracepath`).
+
+## Run with Unikraft
+
+Run the `run.sh` script from the [`run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader). There are four different executables, as listed below:
+
+```console
+./run.sh -r ../dynamic-apps/iputils_srcs/app-iputils /bin/ping
```
Comment (RaduNichita) [inline iputils-src/README.md:35]:
```suggestion
./run.sh -r ../dynamic-apps/iputils_src/app-iputils /bin/ping 8.8.8.8
```

File: iputils-src/README.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+./run.sh -r ../dynamic-apps/iputils_srcs/app-iputils /bin/arping
```
Comment (RaduNichita) [inline iputils-src/README.md:36]:
```suggestion
./run.sh -r ../dynamic-apps/iputils_src/app-iputils /bin/arping 172.44.0.1
```

File: iputils-src/README.md [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+./run.sh -r ../dynamic-apps/iputils_srcs/app-iputils /bin/clockdiff
```
Comment (RaduNichita) [inline iputils-src/README.md:37]:
```suggestion
./run.sh -r ../dynamic-apps/iputils_src/app-iputils /bin/clockdiff 127.0.0.1
```

File: iputils-src/README.md [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+./run.sh -r ../dynamic-apps/iputils_srcs/app-iputils /bin/tracepath
```
Comment (RaduNichita) [inline iputils-src/README.md:38]:
```suggestion
./run.sh -r ../dynamic-apps/iputils_src/app-iputils /bin/tracepath 127.0.0.1
```

**Review by RaduNichita — CHANGES_REQUESTED** [review]

File: iputils-src/build.sh [hunk 15]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+
+rm -rf iputils
+
+# echo "Downloading the source code"
```
Comment (RaduNichita) [inline iputils-src/build.sh:5]: Why are there commented?

**Comment by StefanJum** [comment]
@RaduNichita I removed the installation step.

**Review by RaduNichita — APPROVED** [review]
All good now, thanks @StefanJum and @radupascale :100: 

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

**Review by StefanJum — APPROVED** [review]
Thank you @radupascale :rocket: 
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>
Approved-by: Stefan Jumarea <stefanjumarea02@gmail.com>

**Pull request merged**